xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -747,4 +983,134 @@ private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
+          : null;
+      XdsResourceType.Args args = new XdsResourceType.Args(serverInfo, versionInfo, nonce,
+          bootstrapInfo, securityConfig, toParseResourceNames);
+      handleResourceUpdate(args, resources, xdsResourceType, processingTracker);
+    }
+
+    @Override
+    public void handleStreamClosed(Status status, boolean inRetry,
+                                   ControlPlaneClient cpcThatClosed) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      if (!inRetry) {
+        cleanUpResourceTimers(cpcThatClosed);
+      }
+
+      Status error = status.isOk() ? Status.UNAVAILABLE.withDescription(CLOSED_BY_SERVER) : status;
+      boolean checkForFallback = !status.isOk() && BootstrapperImpl.isEnabledXdsFallback()
+          && !inRetry;
+
+      List<String> authoritiesForClosedCpc = getActiveAuthorities(cpcThatClosed);
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (subscriber.hasResult() || !authoritiesForClosedCpc.contains(subscriber.authority)) {
+            continue;
+          }
+
+          // try to fallback to lower priority control plane client
+          if (checkForFallback && doFallbackForAuthority(
+              cpcThatClosed, serverInfo, subscriber.serverInfos, subscriber.authority)) {
+            authoritiesForClosedCpc.remove(subscriber.authority);
+            if (authoritiesForClosedCpc.isEmpty()) {
+              return; // optimization: no need to continue once all authorities have done fallback
+            }
+            continue;
+          }
+
+          subscriber.onError(error, null);
+        }
+      }
+    }
+
+    @Override
+    public void handleStreamRestarted(ServerInfo serverInfo) {
+      if (isShutDown()) {
+        return;
+      }
+
+      syncContext.throwIfNotInThisSynchronizationContext();
+      ControlPlaneClient controlPlaneClient = serverCpClientMap.get(serverInfo);
+      if (controlPlaneClient == null) {
+        return;
+      }
+
+      for (String authority : getAuthoritiesForServerInfo(serverInfo)) {
+        if (controlPlaneClient.hasSubscribedResources(authority)) {
+          internalHandleStreamReady(serverInfo, controlPlaneClient, authority);
+        }
+      }
+    }
+
+    private void internalHandleStreamReady(
+        ServerInfo serverInfo, ControlPlaneClient controlPlaneClient, String authority) {
+      ControlPlaneClient activeCpClient = getActiveCpc(authority);
+      addCpcToAuthority(authority, controlPlaneClient);
+      if (activeCpClient == null) {
+        restartMatchingSubscriberTimers(authority);
+        controlPlaneClient.sendDiscoveryRequests();
+        return; // Since nothing was active can ignore fallback
+      }
+
+      if (activeCpClient.isReady()
+          && compareCpClients(activeCpClient, controlPlaneClient, authority) < 0) {
```

Comment:
I'm wary of the compare function as it can be hard to notice issues (e.g., is higher priority a bigger number or lower number? Yes that is defined, but you have to think about it), and it seems unnecessary to go through `ServerInfo`.

Can't this just be `if (!activeCpClients.get(authority).contains(controlPlaneClient))`? Note that if you get rid of the call to `getAuthoritiesForServerInfo()` above like I suggest then this condition would disappear.

Reply:
Done